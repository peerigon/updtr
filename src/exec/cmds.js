function installFn(baseCmd) {
    return ({ registry, modules } = {}) =>
        [baseCmd, stringifyRegistry(registry), stringifyModules(modules)].join(
            ""
        );
}

function stringifyModules(modules) {
    return Array.isArray(modules) === true ?
        " " +
              modules
                  // We need to wrap this in double-quotes because some semver
                  // characters like the caret symbol are reserved characters on Windows.
                  .map(({ name, version }) => `"${ name }@${ version }"`)
                  .join(" ") :
        "";
}

function stringifyRegistry(registry) {
    return registry === undefined ? "" : ` --registry "${ registry }"`;
}

function list({ modules } = {}) {
    return [
        "npm ls --json --depth=0",
        Array.isArray(modules) === true ? " " + modules.join(" ") : "",
    ].join("");
}

export default {
    npm: {
        outdated: () => "npm outdated --json --depth=0",
        installMissing: installFn("npm install"),
        install: installFn("npm install"),
        // remove: ({ name }) => ["npm remove ", name].join(""),
        test: () => "npm test",
        list,
    },
    // yarn does not support custom registries yet.
    // However, these renderers accept them anyway.
    yarn: {
        outdated: () => "yarn outdated --json --flat",
        installMissing: installFn("yarn"),
        install: installFn("yarn add"),
        // remove: ({ name }) => ["yarn remove ", name].join(""),
        test: () => "yarn test",
        // We use npm for listing the dependencies because there is no
        // special benefit by using yarn here
        list,
    },
};
