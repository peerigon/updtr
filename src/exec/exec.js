import childProcess from "child_process";

function promiseExec(cwd, cmd) {
    return new Promise(resolve => {
        childProcess.exec(
            cmd,
            { maxBuffer: Infinity, encoding: "utf8", cwd },
            (err, stdout, stderr) => void resolve({ err, stdout, stderr })
        );
    });
}

export default (async function exec(cwd, cmd) {
    const { err, stdout, stderr } = await promiseExec(cwd, cmd);

    if (err !== null) {
        err.stdout = stdout;
        err.stderr = stderr;

        throw err;
    }

    return { stdout, stderr };
});
