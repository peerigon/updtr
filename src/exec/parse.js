const STRING_PROPERTIES = ["name", "current", "wanted", "latest"];

/**
 * Applies JSON.parse on the str.
 * Additionally, returns null if the str was empty or just contained whitespace.
 *
 * @param {string} str
 * @returns {*}
 */
function parse(str) {
    const trimmed = str.trim();

    if (trimmed.length === 0) {
        return null;
    }

    let resultJSON = {};
    try {
        resultJSON = JSON.parse(trimmed);
    } catch (e) {
        const eol = require("os").EOL;
        const lines = trimmed.split(eol);
        lines.map(stringJSON => {
            const possible = JSON.parse(stringJSON);
            if (possible.type === "table") {
                resultJSON = possible;
            }
        })
    }

    return resultJSON;
}

function returnIfValid(result) {
    STRING_PROPERTIES.forEach(prop => {
        if (typeof result[prop] !== "string") {
            throw new Error("Unexpected output format of package manager");
        }
    });

    return result;
}

function tryParse(normalizer) {
    return (stdout, cmd) => {
        try {
            return normalizer(parse(stdout));
        } catch (err) {
            err.message = `Error when trying to parse stdout from command '${ cmd }': ${ err.message }`;
            throw err;
        }
    };
}

function arrToObj(arr, keys) {
    return keys.reduce((obj, key, i) => {
        obj[key] = arr[i];

        return obj;
    }, {});
}

// By sorting the parsed data, we get deterministic results across different npm and yarn versions.
// As a nice side-effect, a package like eslint will always be updated before eslint-config-peerigon
// which might have a peer dependency on eslint
// See https://github.com/peerigon/updtr/issues/48
function sortByName(o1, o2) {
    return o1.name > o2.name;
}

const list = tryParse(
    parsed =>
        (parsed.dependencies === undefined ?
            [] :
            Object.keys(parsed.dependencies)
                .map(name => ({
                    name,
                    version: parsed.dependencies[name].version,
                }))
                .sort(sortByName))
);

export default {
    npm: {
        outdated: tryParse(parsed => {
            if (parsed === null) {
                return [];
            }

            const names = Object.keys(parsed);

            return names
                .map(name => parsed[name])
                .map((dep, index) =>
                    returnIfValid({
                        name: names[index],
                        current: dep.current,
                        wanted: dep.wanted,
                        latest: dep.latest,
                    })
                )
                .sort(sortByName);
        }),
        list,
    },
    yarn: {
        outdated: tryParse(
            parsed =>
                (parsed === null ?
                    [] :
                    parsed.data.body
                        .map(row => arrToObj(row, parsed.data.head))
                        .map(dep =>
                            returnIfValid({
                                name: dep.Package,
                                current: dep.Current,
                                wanted: dep.Wanted,
                                latest: dep.Latest,
                            })
                        )
                        .sort(sortByName))
        ),
        // We currently only use npm for the list command (see cmds.js).
        // Put the real implementation here if we decide to use yarn for list.
        list,
    },
};
