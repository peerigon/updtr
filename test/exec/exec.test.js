import exec from "../../src/exec/exec";

const logMemoryCode = (() => {
    const DEFAULT_STDOUT_BUFFER_SIZE = 200 * 1024;
    const size = DEFAULT_STDOUT_BUFFER_SIZE + 1;
    const memory = typeof Buffer.allocUnsafe === "function" ?
        Buffer.allocUnsafe(size) :
        new Buffer(size);

    console.log(memory.toString("utf8"));
})
    .toString()
    .replace(/"/g, '\\"')
    .replace(/\r?\n/g, "");

const logOkCmd = 'node -e "console.log(\\"ok\\")"';
const noopCmd = 'node -e ""';
const throwCmd = "node -e \"throw 'fail'\"";
const logMemoryCmd = `node -e "(${ logMemoryCode })()"`;
const cwd = __dirname;

describe("exec()", () => {
    it("should return a promise that resolves to an object of expected shape", async () => {
        const result = await exec(cwd, noopCmd);

        expect(result).toMatchSnapshot();
    });
    it("should return a promise that resolves to a value that contains the stdout", async () => {
        const result = await exec(cwd, logOkCmd);

        expect(result.stdout).toBe("ok\n");
    });
    it("should not fail with 'stdout maxBuffer exceeded' if stdout is pretty big", async () => {
        await exec(cwd, logMemoryCmd);
    });
    describe("when the command fails", () => {
        it("should reject the promise with an error of expected shape", async () => {
            let givenErr;

            try {
                await exec(cwd, throwCmd);
            } catch (err) {
                givenErr = err;
            }

            expect(givenErr.message).toContain(throwCmd);
            expect(givenErr).toHaveProperty("stdout");
            expect(givenErr).toHaveProperty("stderr");
            expect(givenErr.stderr).toMatch(/throw 'fail'/);
        });
    });
});
