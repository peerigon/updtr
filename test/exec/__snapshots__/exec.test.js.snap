// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`exec() should return a promise that resolves to an object of expected shape 1`] = `
Object {
  "stderr": "",
  "stdout": "",
}
`;
